lte/gateway/c/core/oai/test/mme_app_task/test_mme_app_emm_encode_decode.cpp

```diff
@@ -0,0 +1,255 @@
+ * limitations under the License.
+ */
+#include <gtest/gtest.h>
+
+extern "C" {
+#include "lte/gateway/c/core/oai/tasks/nas/emm/msg/AttachRequest.h"
+#include "lte/gateway/c/core/oai/tasks/nas/emm/msg/EmmInformation.h"
+#include "lte/gateway/c/core/oai/tasks/nas/emm/msg/TrackingAreaUpdateRequest.h"
+#include "lte/gateway/c/core/oai/tasks/nas/emm/msg/UplinkNasTransport.h"
+#include "lte/gateway/c/core/oai/common/dynamic_memory_check.h"
+#include "lte/gateway/c/core/oai/common/log.h"
+}
+
+#include "lte/gateway/c/core/oai/test/mme_app_task/mme_app_test_util.h"
+
+namespace magma {
+namespace lte {
+
+#define BUFFER_LEN 200
+
+#define FILL_EMM_COMMON_MANDATORY_DEFAULTS(msg)                                \
+  do {                                                                         \
+    msg.protocoldiscriminator = EPS_SESSION_MANAGEMENT_MESSAGE;                \
+    msg.securityheadertype    = 0x0001;                                        \
+    msg.messagetype           = 33;                                            \
+  } while (0)
+
+#define ASSERT_EQ_GUTI_EPS_MOBILE_IDENTITY(orig_guti, dec_guti)                \
+  ASSERT_EQ(orig_guti.guti.mcc_digit1, dec_guti.guti.mcc_digit1);              \
+  ASSERT_EQ(orig_guti.guti.mcc_digit2, dec_guti.guti.mcc_digit2);              \
+  ASSERT_EQ(orig_guti.guti.mcc_digit3, dec_guti.guti.mcc_digit3);              \
+  ASSERT_EQ(orig_guti.guti.mnc_digit1, dec_guti.guti.mnc_digit1);              \
+  ASSERT_EQ(orig_guti.guti.mnc_digit2, dec_guti.guti.mnc_digit2);              \
+  ASSERT_EQ(orig_guti.guti.mnc_digit3, dec_guti.guti.mnc_digit3);              \
+  ASSERT_EQ(orig_guti.guti.mme_code, dec_guti.guti.mme_code);                  \
+  ASSERT_EQ(orig_guti.guti.m_tmsi, dec_guti.guti.m_tmsi);
+
+class EMMEncodeDecodeTest : public ::testing::Test {
+  virtual void SetUp() {}
+  virtual void TearDown() {}
+
+ protected:
+  uint8_t temp_buffer[BUFFER_LEN];
+};
+
+TEST_F(EMMEncodeDecodeTest, TestDecodeAttachRequest1) {
+  //   Combined attach, NAS message generated from s1ap tester
+  uint8_t buffer[] = {0x72, 0x08, 0x09, 0x10, 0x10, 0x00, 0x00, 0x00,
+                      0x00, 0x10, 0x02, 0xe0, 0xe0, 0x00, 0x04, 0x02,
+                      0x01, 0xd0, 0x11, 0x40, 0x08, 0x04, 0x02, 0x60,
+                      0x04, 0x00, 0x02, 0x1c, 0x00};
+  uint32_t len     = 29;
+  attach_request_msg attach_request = {0};
+
+  int decoded = decode_attach_request(&attach_request, buffer, len);
+  ASSERT_EQ(decoded, len);
+  ASSERT_EQ(attach_request.epsattachtype, EPS_ATTACH_TYPE_COMBINED_EPS_IMSI);
+  ASSERT_EQ(
+      attach_request.naskeysetidentifier.naskeysetidentifier,
+      NAS_KEY_SET_IDENTIFIER_NOT_AVAILABLE);
+  ASSERT_EQ(attach_request.naskeysetidentifier.tsc, 0);
+
+  bdestroy_wrapper(&attach_request.esmmessagecontainer);
+  bdestroy_wrapper(&attach_request.supportedcodecs);
+}
+
+TEST_F(EMMEncodeDecodeTest, TestDecodeAttachRequest2) {
```

Comment:
let's also convert this to decode/encode sequence as you did with `TestEncodeDecodeAttachRequest`.

Reply:
doing this results in Fatal as this will trigger calling encode_mobile_station_classmark_3_ie() which unfortunately is not implemented yet

Reply:
I see. Can you create a BE task for these Fatal statements for unimplemented IEs?